Add test-server timing and diagnostic infrastructure

jkiviluoto-nv opened the pull request:
Fixes #9249

Re-applies #9066 with fixes for VK-GL-CTS backwards compatibility.

## Features
- Add executionTimeMs to ExecutionResult for test timing measurement
- Add testCommand to ExecuteToolTestArgs for diagnostic context
- Add Chronometer class for measuring test execution phases
- Enable timing via SLANG_TEST_DIAGNOSTICS=timing environment variable

## Diagnostics
- Add SLANG_TEST_DIAGNOSTICS environment variable for debug output control
- Categories: timing, timing-phases, rpc, fd, pipe, all
- Add RPC-level logging for test-server communication debugging
- Add phase timing output for performance analysis

## Backwards compatibility
- Fix convertArrayToStruct to handle Optional fields in array-style JSON-RPC
- Older clients sending fewer array elements now work if missing fields are Optional
- Enables backward-compatible protocol evolution

jkiviluoto-nv committed:
Add diagnostic logging for slang-test debugging

Fixes #9249

Re-applies #9066 with fixes for VK-GL-CTS backwards compatibility.

Features:
- Add executionTimeMs to ExecutionResult for test timing measurement
- Add testCommand to ExecuteToolTestArgs for diagnostic context
- Add Chronometer class for measuring test execution phases
- Enable timing via SLANG_TEST_DIAGNOSTICS=timing environment variable

Diagnostics:
- Add SLANG_TEST_DIAGNOSTICS environment variable
  for debug output control
- Categories: timing, timing-phases, rpc, fd, pipe, all
- Add RPC-level logging for test-server communication debugging
- Add phase timing output for performance analysis

Backwards compatibility:
- Fix convertArrayToStruct to handle Optional fields in array-style
  JSON-RPC
- Older clients sending fewer array elements now work if missing fields
  are Optional
- Enables backward-compatible protocol evolution

jkwak-work reviewed (commented):
Were you able to figure out exactly what caused VK-CTS to fail?

jkiviluoto-nv commented on source/compiler-core/slang-test-server-protocol.h:
```diff
@@ -7,6 +7,17 @@
+// 2. Then add the new fields here
+//
+// The VK-GL-CTS nightly CI workflow will fail if this ABI is broken.
+
```
The nightly uses VK-GL-CTS release zip, which is built with Slang. That's the reference to prebuilt. And as the old release is built before my diagnostics change, it has incompatible protocol version. That was the root cause of nightly run failure.

https://github.com/KhronosGroup/VK-GL-CTS/compare/main...shader-slang:VK-GL-CTS:main#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20a

jkiviluoto-nv replied:
The VK-GL-CTS release package has the old test-server. Mine diagnostics patch changed the protocol. Then nightly VK-GL-CTS run fails.

jkiviluoto-nv commented:
The reason for failure was root caused - JSON-RPC array parser did not handle optional fields properly. Fixed, squashed the PR, tested with a temporary branch dispatching the VK CTS nightly manually, passes: https://github.com/shader-slang/slang/actions/runs/20089944195

jkwak-work commented on tools/slang-test/slang-test-main.cpp:
```diff
@@ -1142,6 +1185,11 @@ static Result _executeRPC(
+
+    // Drain test-server stderr (phase timing diagnostics) before returning
+    // so it appears before the test result
+    context->drainTestServerStderr();
```
Shouldn't we report the error through
`outRes.standardError` rather than draining it?

jkiviluoto-nv replied:
The diagnostic output is not really an error from the test — it's debug/timing info from the test infrastructure itself (RPC timing, pipe stats, phase timing). Routing it through `outRes.standardError` would mix it with actual test error output and could interfere with test result matching. Draining it directly to stderr keeps it separate. Added a comment explaining the safety of calling this while the process is running (non-blocking pipe reads).

jkiviluoto-nv commented:
@claude please review the feedback and test the PR locally.

mkeshavaNV commented:
@jkiviluoto-nv - Doing a PR cleanup pass. Is this still needed?

jkiviluoto-nv committed:
Address review feedback for diagnostic infrastructure

- Use PlatformUtil::getEnvironmentVariable instead of raw getenv/_dupenv_s
- Use case-insensitive comparison for diagnostic category matching
- Fix SCOPED_TIMER macro: two-level indirection for __LINE__ expansion
- Fix missing (long long) cast in Windows fprintf for RPC write duration
- Remove deprecated slang.dll copy from VK-GL-CTS nightly workflow
- Add comments explaining drainTestServerStderr safety and stderr logging

jkiviluoto-nv commented on source/core/slang-stream.cpp:
```diff
@@ -191,6 +197,28 @@ SlangResult FileStream::_init(
+            fileName.getBuffer(),
+            mode,
+            fileno(m_handle),
+            (unsigned long)pthread_self());
```
Yes, this works on macOS — the `#ifndef _WIN32` path uses POSIX APIs (`pthread_self`, `fileno`) which are available on both Linux and macOS. Verified locally on macOS arm64.

jkiviluoto-nv commented on source/core/slang-test-diagnostics.h:
```diff
@@ -0,0 +1,44 @@
+        static char buffer[256];
+        char* buf = nullptr;
+        size_t len = 0;
+        if (_dupenv_s(&buf, &len, "SLANG_TEST_DIAGNOSTICS") == 0 && buf != nullptr)
```
Good suggestion, switched to `PlatformUtil::getEnvironmentVariable` in the latest commit.
